api/workload/v3/workload.go

```diff
@@ -0,0 +1,218 @@
+	// that the workload should trust, keyed by the SPIFFE ID of the foreign domain.
+	// Unimplemented.
+	FederatedBundles map[string]*x509.CertPool
+}
+
+// Default returns the default SVID (the first in the list).
+//
+// See the SPIFFE Workload API standard Section 5.3
+// (https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE_Workload_API.md#53-default-identity)
+func (x *X509SVIDs) Default() *X509SVID {
+	return x.SVIDs[0]
+}
+
+// SVID is an X.509 SPIFFE Verifiable Identity Document.
+//
+// See https://github.com/spiffe/spiffe/blob/master/standards/X509-SVID.md
+type X509SVID struct {
+	SPIFFEID     string
+	PrivateKey   crypto.Signer
+	Certificates []*x509.Certificate
+	TrustBundle  *x509.CertPool
+}
+
+// WorkloadX509SVIDWatcher is implemented by consumers who wish to be updated on SVID changes.
+type WorkloadX509SVIDWatcher interface {
+	// UpdateX509SVIDs indicates to the Watcher that the SVID has been updated
+	UpdateX509SVIDs(*X509SVIDs)
+
+	// OnError indicates an error occurred.
+	OnError(err error)
+
+	// OnConnection indicates a change in the connection state to the SPIFFE agent.
+	OnConnection(connected bool)
+}
+
+// X509SVIDClient interacts with the SPIFFE Workload API.
+type X509SVIDClient struct {
+	logger   *logrus.Logger
+	watcher  WorkloadX509SVIDWatcher
+	addr     string
+	wg       sync.WaitGroup
+	ctx      context.Context
+	backoff  *backoff
+	cancelFn func()
+}
+
+// Option configures the workload client.
+type Option func(*X509SVIDClient)
+
+// Addr specifies the unix socket address of the SPIFFE agent.
+func Addr(addr string) Option {
+	return func(w *X509SVIDClient) {
+		w.addr = addr
+	}
+}
+
+// Logger specifies the logger to use.
+func Logger(logger *logrus.Logger) Option {
+	return func(w *X509SVIDClient) {
+		w.logger = logger
+	}
+}
+
+// NewX509SVIDClient returns a new Workload API client for X.509 SVIDs.
+func NewX509SVIDClient(watcher WorkloadX509SVIDWatcher, opts ...Option) (*X509SVIDClient, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	c := &X509SVIDClient{
+		logger:   logrus.StandardLogger(),
+		addr:     DefaultAgentAddress,
+		watcher:  watcher,
+		ctx:      ctx,
+		cancelFn: cancel,
+		backoff:  newBackoff(),
+	}
+	for _, opt := range opts {
+		opt(c)
+	}
+	if !strings.HasPrefix(c.addr, _unixPathPrefix) {
+		return nil, fmt.Errorf("spiffe/workload: agent address %q is not a unix address", c.addr)
+	}
+	return c, nil
+}
+
+// Start starts the client.
+//
+// The client will always start, and users should rely on the watcher
+// interface to receives updates on the client's status.
+func (c *X509SVIDClient) Start(ctx context.Context) error {
+	c.wg.Add(1)
+	go c.run()
+	return nil
+}
+
+func (c *X509SVIDClient) run() {
+	defer c.wg.Done()
+	for {
+		if done := c.watch(); done {
+			return
+		}
+	}
+}
+
+// creates single watch loop and returns whether we should continue watching
+func (c *X509SVIDClient) watch() (done bool) {
+	conn, err := c.newConn()
```

Comment:
I've been spending more time reading the GRPC code and I don't think we need to create a new conn here. GRPC should be managing the subconns in the background (in this case a single connection to a unix socket). Creating a new stream will cause it to hit the balancer which will either succeed because there is an active subconn, or fail because there isn't and we'll need to retry around that.

Additionally, given we need to implement that retry logic, we *may* not want grpc.WithBlock. I'll need to go re-read the docs there. We may still want it to enable the client's context to manage the conn.

Reply:
removed WithBlock so that we can implement the initial connection retry ourselves and surface the error to users